steel/docs/how-does-steel-work.md

```diff
@@ -0,0 +1,130 @@
+
+```rust
+// HOST PROGRAM
+// Create an alloy provider from RPC URL
+let provider = ProviderBuilder::new()
+	.with_recommended_fillers()
+	.on_http(args.eth_rpc_url);
+
+// Create an EVM environment from that provider defaulting to the latest block.
+let mut env = EthEvmEnv::builder()
+	.provider(provider.clone())
+	.build()
+	.await?;
+
+// Preflight the call to prepare the input that is required to execute the function in the guest without RPC access. 
+let mut contract = Contract::preflight(args.token_contract, &mut env);
+
+let evm_input = env.into_input().await?
+```
+
+ `preflight` handles calling the RPC provider for the necessary state and for the Merkle storage proofs via `eth_getProof` ([EIP-1186]). These Merkle proofs are given to the guest which verifies them to prove that the RPC data is valid, without having to run a full node and without trusting the host or RPC provider.
+
+## Verifying the proof on-chain
+
+At this point, we have generated a proof of: a view call of state on-chain and some execution based on that view call state (i.e. checking that the balance is at least 1). 
+
+When using Steel, the general pattern for onchain functions incorporating Steel follows this pseudo-code:
+
+```solidity
+contract {
+	function doSomething(journalData, proof) {
+		validate journal data
+		validate Steel commitment
+
+		verify proof
+		doSomethingElse() 
+}
+
+}
+```
+
+The interesting on-chain logic, *doSomethingElse()*, is only reached if the journal data, the steel commitment and the proof are all valid. 
+
+Concretely, in the erc20-counter example, the counter is only updated if the caller has a balance of at least one, and this counter update is gated by Steel and the zkVM.
+
+```solidity
+contract Counter {
+
+function increment(bytes calldata journalData, bytes calldata seal) external {
+        // Decode and validate the journal data
+        Journal memory journal = abi.decode(journalData, (Journal));
+        require(journal.tokenContract == tokenContract, "Invalid token address");
+        require(Steel.validateCommitment(journal.commitment), "Invalid commitment");
+
+        // Verify the proof
+        bytes32 journalHash = sha256(journalData);
+        verifier.verify(seal, imageID, journalHash);
+        
+	 // If the balance is at least one, update the counter
+        counter += 1;
+    }
+```
+
+Within a single proof, we’ve seen Steel can handle hundreds of thousands of view calls. Specifically, one partner application has shown gas savings of 1.2 *billion* gas for a contract call using around 400,000 SLOADs. 1.2 billion gas is around 30 *blocks* worth of execution and this can be verified onchain in one proof, that costs under $10 to generate, and less than 300k gas to verify (see [RISC Zero’s verification contracts]).
```

Comment:
Hundreds of thousands of view calls is slightly different than what we've seen. It should be possible, but is not something we've actually done yet.
```suggestion
Within a single proof, we’ve seen Steel can handle view calls orders of magnitude larger than on-chain execution can handle. Specifically, one partner application has shown gas savings of 1.2 *billion* gas for a contract call using around 400,000 SLOADs. 1.2 billion gas is around 30 *blocks* worth of execution and this can be verified onchain in one proof, that costs under $10 to generate, and less than 300k gas to verify (see [RISC Zero’s verification contracts]).
```